lib/connection/pool.js

```diff
@@ -957,64 +994,118 @@ Pool.prototype.write = function(commands, options, cb) {
     cb: cb, raw: false, promoteLongs: true, promoteValues: true, promoteBuffers: false, fullResult: false
   };
 
-  var buffer = null
+  // Ensure commands is an array
+  if (!Array.isArray(commands)) {
+    commands = [commands];
+  }
 
-  if(Array.isArray(commands)) {
-    buffer = [];
+  // For each command, create a promise while we create the command to send to the server
+  var commandsPromises = []
```

Comment:
This improvement makes sense to me!

However, I think that the method you are proposing will only work if `commands` has length one (see https://github.com/christkv/mongodb-core/blob/2.0/lib/connection/pool.js#L973-L985). If there is more than one command, then each of these commands needs to be individually transformed into an OP_COMPRESSED message, which would require multiple async compression tasks. The only tidy way I can think to handle this is through `Promise.all()`.

That linked section of code from 2.0 appears to suggest that sometimes it is possible for more than one command to be sent at the same time. However, I just ran the tests with a console log and I can't see any moment during the tests when more than one command is sent at the same time.

Do you know whether multiple commands are ever actually sent? If multiple commands _are_ ever sent, how would you advise I best wait for the multiple async compression tasks to complete?